tensorboard/components/plugin_util/plugin-host.ts

```diff
@@ -12,46 +12,87 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-import {IPC, Message, MessageType, PayloadType} from './message.js';
-
-class HostIPC extends IPC {
-  sendMessage(
-    iframe: HTMLIFrameElement,
-    type: MessageType,
-    payload: PayloadType
-  ): Promise<PayloadType> {
-    return this.sendMessageToWindow(iframe.contentWindow, type, payload);
+import {IPC, MessageType, PayloadType, MessageCallback} from './message.js';
+
+const portIPCs = new Set<IPC>();
+const VERSION = 'experimental';
+const listeners = new Map<MessageType, MessageCallback>();
+
+// TODO(@psybuzz): replace this and the port cleanup logic in broadcast() with
+// a MutationObserver to notify us when iframes disconnect.
+const ipcToFrame = new Map<IPC, HTMLIFrameElement>();
+
+// The initial Window-level listener is needed to bootstrap only.
+// All further communication is done over MessagePorts.
+window.addEventListener('message', (event) => {
+  if (event.data !== `${VERSION}.bootstrap`) return;
+  const port = event.ports[0];
+  if (!port) return;
+  const frame = event.source ? event.source.frameElement : null;
+  if (!frame) return;
+
+  const portIPC = new IPC(port);
+  portIPCs.add(portIPC);
+  ipcToFrame.set(portIPC, frame as HTMLIFrameElement);
+  port.start();
+
+  for (const [type, callback] of listeners) {
+    portIPC.listen(type, callback);
+  }
+});
+
+function _broadcast(
+  type: MessageType,
+  payload: PayloadType
+): Promise<PayloadType[]> {
+  for (const ipc of portIPCs) {
+    if (!ipcToFrame.get(ipc).isConnected) {
```

Comment:
I presume this cleanup code can removed with your TODO in L21?

Reply:
Yes!  I can add another comment...